Wikipedia/UI-V5/WMFArticleHeaderImageGalleryViewController.m

```diff
@@ -95,45 +98,141 @@ - (UICollectionViewCell*)collectionView:(UICollectionView*)collectionView
-                    [imageMetadata setNormalizedFaceBoundsFromFeatures:faces inImage:image];
-                    [imageMetadata save];
-                    [self setImage:image centeringBounds:imageMetadata.firstFaceBounds forCellAtIndexPath:indexPath];
-                });
-            } else {
-                [self setImage:image centeringBounds:imageMetadata.firstFaceBounds forCellAtIndexPath:indexPath];
-                return nil;
-            }
+            DDLogVerbose(@"Setting image %@ after retrieving from %@", imageMetadata.sourceURL, download.origin);
+            [self setImage:image
+             inCellAtIndexPath:indexPath
+               centeringBounds:imageMetadata.firstFaceBounds
+                      animated:shouldAnimate];
+            return nil;
         }
+    })
+    .catch(^(NSError* error) {
+        // TODO: show error in UI
+        DDLogError(@"Failed to fetch image from %@. %@", [imageMetadata sourceURL], error);
     });
-    return cell;
 }
 
-- (void)setImage:(UIImage*)image centeringBounds:(CGRect)normalizedCenterBounds forCellAtIndexPath:(NSIndexPath*)path {
-    WMFImageCollectionViewCell* cell = (WMFImageCollectionViewCell*)[self.collectionView cellForItemAtIndexPath:path];
+/**
+ *  Attempt to populate the given `cell` with the image for `metadata` if it's stored in memory.
+ *
+ *  @param cell      The cell to populate.
+ *  @param indexPath The indexPath where the cell resides.
+ *  @param metadata  The metadata which specifies the image URL and face detection information.
+ *
+ *  @return `YES` if the image was set from memory, `NO` if a fetch or other processing needs to take place.
+ */
+- (BOOL)setCachedImageForCell:(WMFImageCollectionViewCell*)cell
+                  atIndexPath:(NSIndexPath*)indexPath
+                     metadata:(MWKImage*)metadata {
+    if (!metadata.didDetectFaces) {
+        return NO;
+    }
+    UIImage* cachedImage = [[WMFImageController sharedInstance] cachedImageInMemoryWithURL:metadata.sourceURL];
+    if (!cachedImage) {
+        return NO;
+    }
+    DDLogVerbose(@"%@ was set at indexPath %@ from memory cache.", metadata.sourceURL, indexPath);
+    [self setImage:cachedImage
+              inCell:cell
+     centeringBounds:metadata.firstFaceBounds
+            animated:NO];
+    return YES;
+}
+
+/**
+ *  Populate the cell at `indexPath` with the given `image`, if it's visible.
+ *
+ *  @param image                  The image to set in the cell's image view.
+ *  @param indexPath              The indexPath for the cell to update.
+ *  @param normalizedCenterBounds The bounds to center in the image view.
+ *  @param animated               Whether or not to animate the transition to `image`.
+ *
+ *  @see setImage:inCell:centeringBounds:animated:
+ */
+- (void)     setImage:(UIImage*)image
+    inCellAtIndexPath:(NSIndexPath*)indexPath
+      centeringBounds:(CGRect)normalizedCenterBounds
+             animated:(BOOL)animated {
```

Comment:
Done really need 3 versions of this method - its kinda confusing. 

For instance, this one is only getting the cell and then calling through to the main method - so its not really saving any typing here. Why not just call these 2 lines where you need them instead of creating another similarly named method?